merge task2 to main

igor-zhuravlev opened the pull request.

igor-zhuravlev committed:
added task2

AliaksandrLiulis commented on task2/rest-api-basics/.gitattributes:
```diff
@@ -0,0 +1,6 @@
+#
```
@igor-zhuravlev extra file

AliaksandrLiulis commented on task2/rest-api-basics/build.gradle:
```diff
@@ -0,0 +1,62 @@
+    }
+
+    dependencies {
+        implementation group: 'org.apache.logging.log4j', name: 'log4j-core', version: '2.14.0'
```
@igor-zhuravlev Could you move all version numbers to const.

AliaksandrLiulis commented on task2/rest-api-basics/domain/src/main/java/com/epam/esm/entity/GiftCertificate.java:
```diff
@@ -0,0 +1,124 @@
+public class GiftCertificate implements Serializable {
+    private static final long serialVersionUID = 7468316931994434280L;
+
+    private Long id;
```
@igor-zhuravlev The same (Certificate, Tag) id and name fields. Could be moved to abstract class.

AliaksandrLiulis commented on task2/rest-api-basics/gradlew:
```diff
@@ -0,0 +1,185 @@
+#!/usr/bin/env sh
```
@igor-zhuravlev extra file

AliaksandrLiulis commented on task2/rest-api-basics/gradlew.bat:
```diff
@@ -0,0 +1,89 @@
+@rem
```
@igor-zhuravlev move all extra files to git ignore

AliaksandrLiulis replied:
@igor-zhuravlev also you have some empty files like task2/rest-api-basics/profile-dev.gradle

AliaksandrLiulis commented on task2/rest-api-basics/repository/build.gradle:
```diff
@@ -0,0 +1,44 @@
+dependencies {
+    implementation project(':domain')
+
+    implementation group: 'org.postgresql', name: 'postgresql', version: '42.2.18'
```
@igor-zhuravlev Could you move all version numbers to const

AliaksandrLiulis commented on task2/rest-api-basics/repository/src/main/java/com/epam/esm/repository/GiftCertificateRepository.java:
```diff
@@ -0,0 +1,26 @@
+    GiftCertificate findByName(String name) throws RepositoryException;
+    List<GiftCertificate> findByTagName(String tagName) throws RepositoryException;
+    Long updateById(Long id, GiftCertificate giftCertificate) throws RepositoryException;
+    Long deleteByName(String name) throws RepositoryException;
```
@igor-zhuravlev can the tags contain the same names? Why not by id?

AliaksandrLiulis commented on task2/rest-api-basics/repository/src/main/java/com/epam/esm/repository/GiftCertificateRepository.java:
```diff
@@ -0,0 +1,26 @@
+    Long updateById(Long id, GiftCertificate giftCertificate) throws RepositoryException;
+    Long deleteByName(String name) throws RepositoryException;
+
+    Set<GiftCertificate> findAllWithTags() throws RepositoryException;
```
@igor-zhuravlev Why do you need so many identical methods? Can you define the execution logic at the service level, generate a request, and call one method? There are a lot of unnecessary methods.

AliaksandrLiulis commented on task2/rest-api-basics/repository/src/main/java/com/epam/esm/repository/impl/GiftCertificateRepositoryImpl.java:
```diff
@@ -0,0 +1,321 @@
+    private static final String FIND_ALL_QUERY =
+            "SELECT id gc_id, name gc_name, description, price, duration, create_date, last_update_date " +
+            "FROM gift_certificate";
+    private static final String FIND_ALL_WITH_TAGS_QUERY =
```
@igor-zhuravlev A lot of copypast.  Your queries are almost the same in terms of parameters and differ only in search and sorting parameters. Why not generate all this in one place?

AliaksandrLiulis commented on task2/rest-api-basics/repository/src/main/java/com/epam/esm/repository/impl/GiftCertificateRepositoryImpl.java:
```diff
@@ -0,0 +1,321 @@
+                    id
+            );
+
+            if (giftCertificate.getTags() != null) {
```
@igor-zhuravlev All logic to service layer.

AliaksandrLiulis commented on task2/rest-api-basics/repository/src/main/java/com/epam/esm/repository/impl/GiftCertificateRepositoryImpl.java:
```diff
@@ -0,0 +1,321 @@
+        try {
+            return jdbcTemplate.queryForObject(FIND_BY_ID_QUERY, new GiftCertificateMapper(), id);
+        } catch (EmptyResultDataAccessException e) {
+            return null;
```
@igor-zhuravlev What does it mean? You can apply ifexist method end return exception with message

mvladmoss replied:
I can suggest using the Optional class for such purposes.

igor-zhuravlev replied:
It throws anyway if not found, thats why I had used catch block

AliaksandrLiulis commented on task2/rest-api-basics/repository/src/main/java/com/epam/esm/repository/impl/GiftCertificateRepositoryImpl.java:
```diff
@@ -0,0 +1,321 @@
+        try {
+            return jdbcTemplate.queryForObject(FIND_BY_NAME_QUERY, new GiftCertificateMapper(), name);
+        } catch (EmptyResultDataAccessException e) {
+            return null;
```
@igor-zhuravlev The same

AliaksandrLiulis commented on task2/rest-api-basics/repository/src/main/java/com/epam/esm/repository/impl/TagRepositoryImpl.java:
```diff
@@ -0,0 +1,99 @@
+    }
+
+    @Override
+    public Long deleteByName(String name) throws RepositoryException {
```
@igor-zhuravlev Why not tag id. It's unique.

AliaksandrLiulis replied:
@igor-zhuravlev if you pass a name from above, just check against the existing name. If exists, return, but not, throw exception

AliaksandrLiulis commented on task2/rest-api-basics/repository/src/main/resources/liquibase/db.changelog-master.xml:
```diff
@@ -0,0 +1,16 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
@extra file

AliaksandrLiulis commented on task2/rest-api-basics/service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java:
```diff
@@ -0,0 +1,230 @@
+
+    @Override
+    public List<GiftCertificateDto> findAllWithTags(Map<String, String[]> params) throws ServiceException {
+        try {
```
@igor-zhuravlev It would be nice to use this construct elsewhere which defines the calling method

AliaksandrLiulis commented on task2/rest-api-basics/web/src/main/java/com/epam/esm/web/controller/GiftCertificateController.java:
```diff
@@ -0,0 +1,101 @@
+    @Autowired
+    private GiftCertificateService giftCertificateService;
+
+    private Map<String, String[]> multiValueMapToMap(MultiValueMap<String, String> requestParams) {
```
@igor-zhuravlev Strange construction.

igor-zhuravlev replied:
It convert request query params to Map<K, List<V>>, because i have the same params like sort (&sort=name,asc&sort=date,desc)

igor-zhuravlev replied:
> @igor-zhuravlev Strange construction.

It convert request params to Map<String, List<>>, because I have the same params like sort (&sort=name,asc&sort=date,desc)

AliaksandrLiulis commented on task2/rest-api-basics/web/src/main/java/com/epam/esm/web/controller/GiftCertificateController.java:
```diff
@@ -0,0 +1,101 @@
+                ));
+    }
+
+    @GetMapping("/gifts-tags")
```
@igor-zhuravlev Would be better underscore

AliaksandrLiulis commented on task2/rest-api-basics/web/src/main/java/com/epam/esm/web/controller/GiftCertificateController.java:
```diff
@@ -0,0 +1,101 @@
+        }
+    }
+
+    @PutMapping("/gifts/{name}")
```
@igor-zhuravlev Would be better by id. Certifictes can be with same names.

AliaksandrLiulis commented on task2/rest-api-basics/web/src/main/java/com/epam/esm/web/controller/GiftCertificateController.java:
```diff
@@ -0,0 +1,101 @@
+        }
+    }
+
+    @DeleteMapping("/gifts/{name}")
```
@igor-zhuravlev id

mvladmoss commented on task2/rest-api-basics/domain/src/main/java/com/epam/esm/entity/GiftCertificate.java:
```diff
@@ -0,0 +1,124 @@
+    private String description;
+    private BigDecimal price;
+    private Integer duration;
+    private LocalDateTime createDate;
```
Suggest to use Instant instead of LocalDateTime

mvladmoss commented on task2/rest-api-basics/repository/src/main/java/com/epam/esm/repository/impl/mapper/GiftCertificateMapper.java:
```diff
@@ -0,0 +1,40 @@
+    public static final String LAST_UPDATE_DATE = "last_update_date";
+
+    public static final String SECONDARY_ID = "id";
+    public static final String SECONDARY_NAME = "name";
```
unused field

mvladmoss commented on task2/rest-api-basics/repository/src/main/java/com/epam/esm/repository/impl/mapper/TagMapper.java:
```diff
@@ -0,0 +1,24 @@
+    public static final String NAME = "t_name";
+
+    public static final String SECONDARY_ID = "id";
+    public static final String SECONDARY_NAME = "name";
```
unused field

mvladmoss commented on task2/rest-api-basics/repository/src/main/java/com/epam/esm/repository/impl/GiftCertificateRepositoryImpl.java:
```diff
@@ -0,0 +1,321 @@
+
+@Repository
+public class GiftCertificateRepositoryImpl implements GiftCertificateRepository {
+    private static final Logger logger = LogManager.getLogger(GiftCertificateRepositoryImpl.class);
```
The logger is declared but is not used anywhere

mvladmoss commented on task2/rest-api-basics/repository/src/main/java/com/epam/esm/repository/impl/GiftCertificateRepositoryImpl.java:
```diff
@@ -0,0 +1,321 @@
+    }
+
+    @Override
+    public Long updateById(Long id, GiftCertificate giftCertificate) throws RepositoryException {
```
Firstly, I suggest renaming it to just "update". Secondly, why do you pass both id and certificate? As I see your GiftCertificate class includes the id field.

igor-zhuravlev replied:
fixed

igor-zhuravlev committed:
edit spring configuration structure on each layer

extend Tag and GiftCertificate classes by AbstractEntity class

update gradle builds

mvladmoss commented on task2/rest-api-basics/service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java:
```diff
@@ -0,0 +1,230 @@
+import java.util.stream.Collectors;
+
+@Service
+public class GiftCertificateServiceImpl implements GiftCertificateService {
```
Please use @Transactiona(readOnly = true) annotation for read operations too.
![image](https://user-images.githubusercontent.com/32880986/104951425-84728000-59d3-11eb-9a99-709ae0991f0c.png)
https://stackoverflow.com/questions/44984781/what-are-advantages-of-using-transactionalreadonly-true

igor-zhuravlev replied:
thanks

mvladmoss commented on task2/rest-api-basics/web/src/main/java/com/epam/esm/web/config/WebAppConfig.java:
```diff
@@ -0,0 +1,73 @@
+    }
+
+    @Bean
+    public DataSource dataSource() {
```
What does configuration which is related to the database do in WebConfig?

mvladmoss commented on task2/rest-api-basics/web/src/main/java/com/epam/esm/web/config/WebAppConfig.java:
```diff
@@ -0,0 +1,73 @@
+import javax.sql.DataSource;
+
+@Configuration
+@EnableTransactionManagement
```
Each of your modules(except domain) should include its own **Configuration** class. Each of these configurations has to make component scan only for this module and customize appropriate beans. In order to include configuration from one module to another one, use annotation @Import.

igor-zhuravlev replied:
thanks

mvladmoss commented on task2/rest-api-basics/web/src/main/java/com/epam/esm/web/config/WebAppConfig.java:
```diff
@@ -0,0 +1,73 @@
+
+    @Bean
+    public PlatformTransactionManager platformTransactionManager() {
+        JdbcTransactionManager jdbcTransactionManager = new JdbcTransactionManager(hikariDataSource());
```
inline here and beyond

igor-zhuravlev replied:
fixed

mvladmoss commented on task2/rest-api-basics/web/src/main/java/com/epam/esm/web/controller/GiftCertificateController.java:
```diff
@@ -0,0 +1,101 @@
+    }
+
+
+    @GetMapping("/gifts")
```
you can put this prefix on the class level because it's duplicated in all endpoints

igor-zhuravlev replied:
fixed

mvladmoss commented on task2/rest-api-basics/web/src/main/java/com/epam/esm/web/controller/GiftCertificateController.java:
```diff
@@ -0,0 +1,101 @@
+    }
+
+    @GetMapping("/gifts-tags")
+    public List<GiftCertificateDto> findAllGiftCertificatesWithTags(@RequestParam(required = false) MultiValueMap<String, String> requestParams) {
```
findAllGiftCertificatesWithTags and findAllGiftCertificates, what is the reason to divide these endpoints?

igor-zhuravlev replied:
fixed

mvladmoss commented on task2/rest-api-basics/web/src/main/java/com/epam/esm/web/controller/GiftCertificateController.java:
```diff
@@ -0,0 +1,101 @@
+        try {
+            return giftCertificateService.findAllWithTags(multiValueMapToMap(requestParams));
+        } catch (ServiceException e) {
+            logger.error(e);
```
For this purpose, you created ExceptionHandler, move logging there

igor-zhuravlev replied:
fixed

mvladmoss commented on task2/rest-api-basics/web/src/main/java/com/epam/esm/web/controller/GiftCertificateController.java:
```diff
@@ -0,0 +1,101 @@
+            return giftCertificateService.findAllWithTags(multiValueMapToMap(requestParams));
+        } catch (ServiceException e) {
+            logger.error(e);
+            throw e;
```
You rethrow ServiceException but it's not handled in ExceptionHandler.

igor-zhuravlev replied:
fixed

igor-zhuravlev committed:
minor fixes

mvladmoss commented:
Igor, why did you resolve discussions? I don't see any commits that fixed the aforementioned remarks or your comments under these remarks.

igor-zhuravlev commented:
> Igor, why did you resolve discussions? I don't see any commits that fixed the aforementioned remarks or your comments under these remarks.

Sorry, I collapsed they for myself readability

mvladmoss commented:
> > Igor, why did you resolve discussions? I don't see any commits that fixed the aforementioned remarks or your comments under these remarks.
> 
> Sorry, I collapsed they for myself readability

Please don't do this

igor-zhuravlev committed:
update abstraction in service and repository layers

add Critera class in repository

add queryBuilder method in QueryUtil

delete tests

mvladmoss commented on task2/rest-api-basics/repository/src/test/java/com/epam/esm/repository/config/RepositoryTestConfig.java:
```diff
@@ -0,0 +1,72 @@
+@Configuration
+@EnableTransactionManagement
+@PropertySource("classpath:database.properties")
+@ComponentScan("com.epam.esm.repository")
```
I suppose if you try to run your test with this config you will face something like NotUniqueBeanException. In this configuration, you created DataSource bean and simultaneously scan the repository package where another Repository configuration with its own DataSource bean exists.

igor-zhuravlev replied:
Thanks, it fixed.

igor-zhuravlev committed:
update find method with criteria in repository

add CriteriaSearch enum

igor-zhuravlev committed:
add user api javadoc

igor-zhuravlev committed:
use Instant instead LocalDateTime

update mappers and convertors as components

igor-zhuravlev committed:
add dto and request params validation

add validation exceptions

minor fixes

igor-zhuravlev committed:
add service unit tests

add dto, params validation in service

igor-zhuravlev committed:
added integration tests

edit init Instant in gift certificate service with truncate in macros for correct
saving Timestamp in database for jdk11

update repository test configuration for correct init DataSource bean

mvladmoss commented on task2/rest-api-basics/repository/src/main/java/com/epam/esm/repository/impl/mapper/TagMapper.java:
```diff
@@ -0,0 +1,25 @@
+    public static final String ID = "t_id";
+    public static final String NAME = "t_name";
+
+    public static final String SECONDARY_ID = "id";
```
Delete unused fields everywhere

igor-zhuravlev replied:
> Delete unused fields everywhere

But it is used in https://github.com/igor-zhuravlev/mjc/blob/c9cb85a679e210b42e235310c95177045a8b78c6/task2/rest-api-basics/repository/src/main/java/com/epam/esm/repository/impl/TagRepositoryImpl.java#L74 when I recieve a generated id

mvladmoss replied:
I didn't notice. Resolved

igor-zhuravlev replied:
> I didn't notice. Resolved

Sorry for strange question, but how I can resolve other conversations? I need a commented their as fixed or maybe simply click resolve

mvladmoss replied:
just resolve them

igor-zhuravlev committed:
separate query builder on chain of Where and Order builders

nikskonda commented on task2/rest-api-basics/service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java:
```diff
@@ -0,0 +1,181 @@
+
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+import java.util.*;
```
How many classes from java.util do you use?

nikskonda commented on task2/rest-api-basics/service/src/test/java/com/epam/esm/service/impl/TagServiceImplTest.java:
```diff
@@ -0,0 +1,325 @@
+    }
+
+    @Test
+    void save_ThrowTagNotValidException() {
```
Just FYI https://dzone.com/articles/7-popular-unit-test-naming

nikskonda commented on task2/rest-api-basics/web/src/main/java/com/epam/esm/web/controller/TagController.java:
```diff
@@ -0,0 +1,59 @@
+     */
+
+    @PostMapping
+    public TagDto saveTag(@RequestBody TagDto tagDto) {
```
I think its better to use "save", "find", ... method name, because its obvious that it will be tag from the name of class.

nikskonda commented on task2/rest-api-basics/service/src/main/java/com/epam/esm/service/constant/ServiceError.java:
```diff
@@ -0,0 +1,43 @@
+        return code;
+    }
+
+    private static class Constants {
```
Why did you move error codes to a separate class?

igor-zhuravlev committed:
apply unit test naming conventions

move codes in ServiceError, delete subclass Codes

minor fixes

nikskonda commented on task2/rest-api-basics/repository/src/main/java/com/epam/esm/repository/config/RepositoryConfig.java:
```diff
@@ -27,7 +27,7 @@ public JdbcTemplate jdbcTemplate() {
     }
 
-    @Bean
+    @Bean(destroyMethod = "close")
```
What is the reason for using destroyMethod = "close"?

igor-zhuravlev replied:
> What is the reason for using destroyMethod = "close"?

I thought if Hikari not autoclosable that put close on bean. Should I delete this?

nikskonda replied:
As I understand, close or shutdown methods will be automatically marked as destroyMethod.
![image](https://user-images.githubusercontent.com/24493000/105823311-0c591b00-5fce-11eb-90a1-4d00bbd95aac.png)

igor-zhuravlev replied:
> As I understand, close or shutdown methods will be automatically marked as destroyMethod.
> ![image](https://user-images.githubusercontent.com/24493000/105823311-0c591b00-5fce-11eb-90a1-4d00bbd95aac.png)

Thank you for your attention. I will fix it.
